Documentation for running a gRPC service on Kubernetes (#1127)
akka/akka-grpc: docs/src/main/paradox/server/kubernetes.md

# Kubernetes

An Akka gRPC service can be deployed to a Kubernetes cluster just like any other HTTP-based application,
but if you want to make sure HTTP/2 is used across the board there are some things to consider.

There are 3 ways to use HTTP/2:

> * With TLS
> * Without TLS ('h2c'), using the `Upgrade` mechanism to negotiate between using HTTP/1 and HTTP/2
> * Without TLS ('h2c'), without negotiation (which assumes the client has prior knowledge that the server supports HTTP/2)
> 

Reviewer: ```suggestion
* HTTP/2 over TLS
* HTTP/2 over plaintext HTTP ('h2c') with negotiation using the `Upgrade` mechanism to upgrade from HTTP/1 to HTTP/2
* HTTP/2 over plaintext HTTP ('h2c') without negotiation which assumes the client has prior knowledge that the server supports HTTP/2

```

I'm not entirely sure about the parens VS comma on the last sentence.

Kubternetes supports many types of [Ingress Controllers](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/),
but [GCE](https://git.k8s.io/ingress-gce/README.md) and [nginx](https://git.k8s.io/ingress-nginx/README.md)
seem to be the most widely known.

> ## LoadBalancer Service

Author: Added it there

By creating a service of type `LoadBalancer` you can expose your services using a direct TCP loadbalancer.
This means all features supported by the underlying implementation are available.

Limitations of this approach are that you are restricted to 1 service per IP, and
more advanced features such as TLS termination and path-based routing are not available.

> ## Nginx Ingress

Reviewer: I stand corrected. The preferred way seems to be all caps (`NGINX`) ass seen in nginx.com (as opposed to nginx.org).

> Nginx, by default, will only support h2c via the `Upgrade` mechanism. Because this mechanism is more
> complicated than the other 2 approaches, client support for this mechanism is not ubiquitous. Also, the
> loadbalancer will use HTTP/1 rather than HTTP/2 to the service. While this may happen to work it is
> not optimal.
> 

Reviewer: ```suggestion
Of the three ways to use HTTP/2 presented at the beginning of this page, nginx only supports 
`h2c` with negotiation via the `Upgrade` mechanism. Because the `Upgrade` mechanism is more
complicated than the other 2 approaches (aka "TLS", and "prior knowledge"), client support 
for `Upgrade` mechanism is not ubiquitous. Also, the loadbalancer will use HTTP/1 rather 
than HTTP/2 to the service. While this may happen to work it is not optimal.

```

I tried to remove as many pronouns as possible. But the current writing seems to focus too much in some details and doesn't paint `nginx` as a very valid option. I liked the `LoadBalancer Service` section because it listed the disadvantages very clearly.

--------------------

I'm not sure how to rewrite the last bit. 

> Also, the loadbalancer will use HTTP/1 rather than HTTP/2 to the service. While this may happen to work it is not optimal.

## GCE Ingress

GCE only supports HTTP/2 over TLS. This means you must configure GCE to use HTTPS both on the
'outside' and on this 'inside', and you will have to make sure your service is exposed over HTTPS

> internally as well.

Reviewer: Hmm, this section also starts indicating what's supported and what's not. Maybe have a repeated list of what's supported or not or what's the default setting, etc... would be helpful.

Or even move that table to the top of the page:

|   | nginx | GCE |
| --------------- | --------------- | --------------- |
| HTTP/2 over TLS | supported | supported |
| h2c + `Upgrade` | default | not supported |
| h2c + prior knowledge | supported | not supported |

Author: While I see the desire for such a table, I'm not sure we can pull it off: all the variations come with their own "ifs and buts", and it is not clear whether this is talking about the connection between the client and the loadbalancer, or the connection between the loadbalancer and the service (or both). Might be interesting to revisit, but I'd like to keep it outside the scope of this PR :).

To make sure the GCE loadbalancer uses HTTP/2 to connect to your service you must
[use the `cloud.google.com/app-protocols` annotation](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress-xlb#https_tls_between_load_balancer_and_your_application) the service:

```
apiVersion: v1
kind: Service
metadata:
  name: my-service-3
  annotations:
    cloud.google.com/app-protocols: '{"my-grpc-port":"HTTP2"}'
spec:
  type: NodePort
  selector:
    app: metrics
    department: sales
  ports:
  - name: my-grpc-port
    port: 443
    targetPort: 8443
```

> By default the loadbalancer will accept any certificate, so your service can be configured with

Reviewer: Sorry, I didn't review this last section

are the most widely known.

## LoadBalancer Service

By creating a service of type `LoadBalancer` you can expose your services using a direct TCP loadbalancer.
This means all features supported by the underlying implementation are available.

Using a `LoadBalancer` as the Ingress has some limitations, for example:

* you are restricted to 1 service per IP
* more advanced features such as TLS termination and path-based routing are not available

## NGINX Ingress

Of the three ways to use HTTP/2 presented at the beginning of this page, NGINX by default only supports 
`h2c` with negotiation via the `Upgrade` mechanism. Because the `Upgrade` mechanism is more
complicated than the other 2 approaches ("TLS", and h2c with "prior knowledge"), gRPC client support 
for h2c with the `Upgrade` mechanism is not ubiquitous. Also, the loadbalancer will use HTTP/1 rather 
than HTTP/2 for the connection from the loadbalancer to the service. While this may happen to work it is not optimal.

## GCE Ingress

GCE only supports HTTP/2 over TLS. This means you must configure GCE to use HTTPS both on the

> 'outside' and on this 'inside', and you will have to make sure your service is exposed over HTTPS

Reviewer: ```suggestion
'outside' and on the 'inside', and you will have to make sure your service is exposed over HTTPS
```
